### PR TITLE
Add weighted image distribution

### PR DESCRIPTION
**Pull request by ClifHouck**
Need to do some functional testing, and probably some logging around choose_weighted_images_forced_distribution, so this pr is just for reviews for the moment.

### CODE REVIEW

File: arsenal/strategy/base.py [hunk 1]
```diff
@@ -265,3 +277,119 @@ def build_attribute_dict(items, attr_name):
...
+            weight = default_weight
+        weights_by_name[name] = weight
+    return weights_by_name
+
+
+def _determine_image_distribution(nodes):
+    """Finds the current distribution of cached images across available nodes.
+
+    Returns a dictionary where the keys are image uuids and the values are
+    the intregral frequency of their occurance.
```
Comment (supermari0) [inline arsenal/strategy/base.py]: s/intregral/integral

File: arsenal/strategy/base.py [hunk 2]
```diff
@@ -265,3 +277,119 @@ def build_attribute_dict(items, attr_name):
...
+
+def get_default_image_weight():
+    """Get the default weight to use for images."""
+    if get_default_image_weight.DEFAULT_IMAGE_WEIGHT is None:
+        if len(CONF.strategy.image_weights.keys()) == 0:
+            get_default_image_weight.DEFAULT_IMAGE_WEIGHT = 1
+        else:
+            # This takes the average of supplied weights and sets that
+            # as the default weight.
+            weight_sum = sum([
```
Comment (supermari0) [inline arsenal/strategy/base.py]: What's the benefit to doing this instead of just having a set default value?

File: arsenal/strategy/base.py [hunk 3]
```diff
@@ -265,3 +277,119 @@ def build_attribute_dict(items, attr_name):
...
+def choose_weighted_images_forced_distribution(num_images, images, nodes):
+    """Returns a list of images to cache
+
+    Factors in the current distribution of images cached across nodes.
+    Enforces the distribution of images to match the weighted distribution as
+    closely as possible.
+
+    It is important to note that there may be circumstances which prevent this
+    function from attaining the desired ideal distribution, but the function
+    will always try its best to reach the desired distirbution based on the
```
Comment (supermari0) [inline arsenal/strategy/base.py]: s/distirbution/distribution

File: arsenal/strategy/base.py [hunk 4]
```diff
@@ -265,3 +277,119 @@ def build_attribute_dict(items, attr_name):
...
+    # the cache.
+    weight_sum = sum([weights_by_name[image.name] for image in images])
+
+    num_cached_nodes = len([node for node in nodes if
+                            node.cached and not node.provisioned])
+    total_desired_cached = num_cached_nodes + num_images
+
+    scale_factor = 1
+    if weight_sum != 0:
+        scale_factor = total_desired_cached / weight_sum
```
Comment (supermari0) [inline arsenal/strategy/base.py]: Won't we have to cast total_desired_cached or weight_sum to a float before doing the division?

File: arsenal/strategy/base.py (hunk 2 above)
Comment (ClifHouck) [inline arsenal/strategy/base.py]: Thinking about this makes me admit I don't have a good justification for taking the average, and using that as a default. I'm going to create a configuration option for it and have the option default to 1. That should be much simpler in the long run.

File: arsenal/strategy/base.py (hunk 4 above)
Comment (ClifHouck) [inline arsenal/strategy/base.py]: You're right that the division operator in Python 2.x defaults to integer division when fed integers. However, added an import for division from the __future__ module, which causes the division operator to always return a floating point number.

File: arsenal/strategy/base.py (hunk 1 above)
Comment (ClifHouck) [inline arsenal/strategy/base.py]: Fixed.

File: arsenal/strategy/base.py (hunk 3 above)
Comment (ClifHouck) [inline arsenal/strategy/base.py]: Fixed/

**Comment by supermari0** [comment]
LGTM but I'd wait for one more +1 before merging

**Pull request merged**